### PR TITLE
Support extracting attributes from the parent namespace

### PR DESCRIPTION
**Pull request by samuelcolvin**
Working on workaround for problems with python type hints namespaces, see #2678.

I'm not yet sure if this is a good idea, it's still confusing since intermediate parent namespaces are ignored.

Still this is an interesting idea.

Will need to be rebased to main once #4516 is merged.

### CODE REVIEW

File: tests/test_forward_ref.py [hunk 1]
```diff
@@ -749,3 +749,94 @@ class Foobar(BaseModel):
...
+    def module():
+        from annotated_types import Gt
+        from typing_extensions import Annotated
+
+        from pydantic import BaseModel
+
+        Foobar = Annotated[int, Gt(0)]  # noqa: F841
+
+        def nested():
+            Foobar = Annotated[int, Gt(10)]
```
Comment (hramezani) [inline tests/test_forward_ref.py:810]:
```suggestion
            Foobar = Annotated[int, Gt(10)]  # noqa: F841
```

Comment (hramezani) [reply tests/test_forward_ref.py]: @samuelcolvin I've just fixed the `CI / lint` problem by merging this.
Now it is green again.

Comment (samuelcolvin) [reply tests/test_forward_ref.py]: great, thanks.

**Comment by samuelcolvin** [comment]
might need #4735

**Pull request merged**